### PR TITLE
[medium] Support fetch of instance meta-data for agents running in AWS

### PR DESCRIPTION
**Pull request by ameihm0912**
This change adds support for discovery of specific elements of instance metadata by agents running in AWS. Fetched metadata is included in the agents environment.

### CODE REVIEW

File: agent.go [hunk 1]
```diff
@@ -36,14 +36,18 @@ type Agent struct {
...
+	OS              string   `json:"os,omitempty"`
+	Arch            string   `json:"arch,omitempty"`
+	IsProxied       bool     `json:"isproxied"`
+	Proxy           string   `json:"proxy,omitempty"`
+	Addresses       []string `json:"addresses,omitempty"`
+	PublicIP        string   `json:"publicip,omitempty"`
+	AWSInstanceID   string   `json:"aws_instanceid,omitempty"`
+	AWSLocalIPV4    string   `json:"aws_localipv4,omitempty"`
+	AWSAMIID        string   `json:"aws_amiid,omitempty"`
+	AWSInstanceType string   `json:"aws_instancetype,omitempty"`
```
Comment (jvehent) [inline agent.go]: I'd prefer having a separate AWS struct that contains a bunch of instance information, as opposed to adding them directly into the Env struct. We could always query them using `agents.environment->'aws'->>'instanceid'`.

File: agent.go [hunk 2]
```diff
@@ -36,14 +36,18 @@ type Agent struct {
...
+	Init            string   `json:"init,omitempty"`
+	Ident           string   `json:"ident,omitempty"`
+	OS              string   `json:"os,omitempty"`
+	Arch            string   `json:"arch,omitempty"`
+	IsProxied       bool     `json:"isproxied"`
+	Proxy           string   `json:"proxy,omitempty"`
+	Addresses       []string `json:"addresses,omitempty"`
+	PublicIP        string   `json:"publicip,omitempty"`
+	AWSInstanceID   string   `json:"aws_instanceid,omitempty"`
+	AWSLocalIPV4    string   `json:"aws_localipv4,omitempty"`
```
Comment (jvehent) [inline agent.go]: Is this different from the content of `addresses` above? If yes, can it be merged into the addresses?

File: mig-agent/aws.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		Dial: (&net.Dialer{Timeout: time.Second}).Dial,
+	}
+	client := &http.Client{Transport: tr}
+	resp, err := client.Get(awsMetaURL() + endpoint)
+	if err != nil {
+		return
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("invalid HTTP response code returned by metadata service")
```
Comment (jvehent) [inline mig-agent/aws.go]: maybe log the response code here, for debugging?

**Comment by jvehent** [comment]
Looks good to me. I was looking for a way to just retrieve a whole JSON from the meta-data in one call, but there doesn't seem to be one. Failing that, you approach is good, I'd just change the struct to put everything into an `AWS` struct.

/me likes :+1:

**Comment by ameihm0912** [comment]
Yeah, I also looked for a similar interface to just get all the meta-data, but there doesn't seem to be anything. It would also be nice if there was a way to obtain instance tags through the same meta-data calls, but this doesn't seem to be possible either.

With the instance-id however, we should be able to cross reference the instance with data obtained from the AWS API that describes the instances, so we can add tags if we need to. This could be done post-request (e.g. as part of a mig-runner processing job).

The suggested changes make sense, I'll take a look at updating the PR.

File: agent.go [hunk 4]
```diff
@@ -36,14 +36,23 @@ type Agent struct {
...
+	PublicIP  string      `json:"publicip,omitempty"`
+	AWS       AgentEnvAWS `json:"aws,omitempty"`
+}
+
+// Stores AWS specific agent environment values
+type AgentEnvAWS struct {
+	AWSInstanceID   string `json:"aws_instanceid,omitempty"`
+	AWSLocalIPV4    string `json:"aws_localipv4,omitempty"`
+	AWSAMIID        string `json:"aws_amiid,omitempty"`
+	AWSInstanceType string `json:"aws_instancetype,omitempty"`
```
Comment (jvehent) [inline agent.go]: I'd drop the `aws_` prefixes in the variables and json, we're already under the `aws` key.

**Comment by jvehent** [comment]
Just one nitpick and it's ready to merge.

**Pull request merged**